### PR TITLE
Refactor p2p layer

### PR DESCRIPTION
**Pull request by LukasGasior1**
## Title
Refactor p2p layer

## Description

Introduce encrypted, RLPx-like peer to peer communication.

Dependent PR - https://github.com/radixdlt/radixdlt-iac/pull/139  and should be merged at the same time

### CODE REVIEW

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/keys/PersistedBFTKeyManager.java [hunk 1]
```diff
@@ -68,4 +68,8 @@ public ECDSASignature sign(byte[] hash) {
 	public BFTNode self() {
 		return BFTNode.create(ecKeyPair.getPublicKey());
 	}
+
+	public ECKeyPair getKeyPair() {
```
Comment (talekhinezh) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/keys/PersistedBFTKeyManager.java:72]: I don't think we should be exposing the private key in this way, better would be if exposed a Signer instead.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/keys/PersistedBFTKeyModule.java [hunk 2]
```diff
@@ -51,6 +52,12 @@ ECPublicKey key(@Self BFTNode bftNode) {
 		return bftNode.getKey();
 	}
 
+	@Provides
+	@Self
+	ECKeyPair ecKeyPair(PersistedBFTKeyManager keyManager) {
```
Comment (talekhinezh) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/keys/PersistedBFTKeyModule.java:57]: I don't think we should be exposing the private key in this way, better would be if exposed a Signer instead.

Comment (talekhinezh) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/keys/PersistedBFTKeyModule.java:57]: E.g. the interface we would have if we were to use a private key in hardware

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/P2PModule.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	@Provides
+	public P2PConfig p2pConfig() {
+		return P2PConfig.fromRuntimeProperties(this.properties);
+	}
+
+	@Provides
+	@Self
+	public RadixNodeUri selfUri(@Self ECPublicKey selfKey, HostIp hostIp, P2PConfig p2pConfig) {
```
Comment (talekhinezh) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/P2PModule.java:96]: Not a fan of having to provide a whole `P2PConfig` to just get the port which is why I like annotations to single values better. But this is fine...for now.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/PeerEvent.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.network.p2p;
+
+import com.radixdlt.network.p2p.transport.PeerChannel;
+
+import java.util.Objects;
+
+public interface PeerEvent {
```
Comment (talekhinezh) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/PeerEvent.java:24]: Some javadoc would be helpful, what's the purpose of this marker interface? Why not just treat each PeerEvent as a separate class without a shared marker interface?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/RadixNodeUri.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+		return nodeId;
+	}
+
+	@JsonValue
+	private byte[] getSerializedValue() {
+		return getUriString().getBytes(StandardCharsets.UTF_8);
+	}
+
+	private String getUriString() {
+		return String.format("radix://%s@%s:%s", nodeId.getPublicKey().toBase58(), host, port);
```
Comment (talekhinezh) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/RadixNodeUri.java:79]: Like mentioned lets use bech32

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/addressbook/AddressBook.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+/**
+ * Manages known peers network addresses and their metadata.
+ */
+@Singleton
+public final class AddressBook {
+	private static final Comparator<AddressBookEntry.PeerAddressEntry> entryComparator = (a, b) -> {
+		final var aLastSuccess = a.getLastSuccessfulConnection().orElse(Instant.MIN);
+		final var bLastSuccess = b.getLastSuccessfulConnection().orElse(Instant.MIN);
+		return -aLastSuccess.compareTo(bLastSuccess);
```
Comment (talekhinezh) [inline radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/addressbook/AddressBook.java:46]: Seems like this comparator basically ignores old entries? Or is there a way to get back to older entries?

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/PeerEvent.java (hunk 4 above)
Comment (LukasGasior1) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/PeerEvent.java:24]: just so we can group related events into a single event dispatcher/processor

**Review by michalzajda — APPROVED** [review]

**Pull request merged**